doc/ecmp/wecmp_hld.md

```diff
@@ -0,0 +1,174 @@
+    "weight":{{weight_list}} (OPTIONAL)
+
+; Defines schema for IPv4/IPv6 route table attributes
+key                         = ROUTE_TABLE:prefix       ; IPv4/IPv6 prefix
+; field                     = value
+nexthop                     = nexthop_list             ; List of nexthops.
+ifname                      = ifname_list              ; List of interfaces.
+weight                      = weight_list              ; List of weights.
+;value annotations
+ifname                      = 1*64VCHAR                ; name of the Interface (Port Channel)
+```
+#### Software Modules
+This section describes modifications to software modules to support Weighted ECMP routes.
+
+##### FPM Syncd
+FPM Syncd is an existing daemon in BGP container that monitors NetLink messages from the SONiC routing stack.
+FPM Syncd now additionally processes Next Hop information in the NetLink messages and sets this information in the APPL DB.
+
+###### fpmsyncd functions
+The following are new functions for fpmsyncd:
+```
+  /* Handler to process ECMP Weight information rtnl messages */
+  string RouteSync::getNextHopWt(struct rtnl_route *route_obj);
+```
+##### RouteOrch
+RouteOrch is an existing component of the OrchAgent daemon in the SWSS container.  RouteOrch monitors operations on Route related tables in APPL DB and converts those operations in SAI commands to manage Route entries.  Additionally RouteOrch coordinates Next Hop object operations with NeighOrch and converts operations into SAI commands to manage Next Hop Group objects.
+
+###### routeorch functions
+The following functions are modified to support weight field in Next Hop Group Members.
+```
+  /* Handler to process ECMP Weight information from APPL DB */
+  void RouteOrch::doPrefixTask(Consumer& consumer)
+  /* Handler to call SAI API with Weight information */
+  bool RouteOrch::validnexthopinNextHopGroup(const NextHopKey &nexthop)
+```
+##### NextHopKey
+NextHopKey is an existing structure of the OrchAgent daemon in the SWSS container.  NextHopKey is used by both RouteOrch and NeighOrch to coordinate Next Hop operations.
+NextHopKey class is modified to include next hop weight field.
+```
+struct NextHopKey
+{
+    IpAddress           ip_address;     // neighbor IP address
+    string              alias;          // incoming interface alias
+    uint8_t             weight;         // NH weight for NHGs
+    ... struct definition abbreviated ...
+};
+```
+
+### SAI API
+This section describes SAI APIs used to support weighted ECMP routes.
+
+Full details about SAI Next Hop Group API and attributes are described here:
+https://github.com/opencomputeproject/SAI/blob/master/inc/sainexthopgroup.h
+To facilitate weighted ECMP route functionality, the following update was made to the SAI Next Hop Group API:
+```
+    /**
+     * @brief Member weights
+     *
+     * @type sai_uint32_t
+     * @flags CREATE_AND_SET
+     * @default 1
+     */
+    SAI_NEXT_HOP_GROUP_MEMBER_ATTR_WEIGHT,

```

Comment:
what is the sai does not support this attribute, what would be the system behavior?

Reply:
When vendor's SAI don't support, it will be ECMP vs WECMP.

Reply:
Check when SAI is supported and handling it accordingly.